File: auth-service/src/main/java/com/bok/auth/service/AuthService.java
package com.bok.auth.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;

import com.bok.auth.entity.RefreshToken;
import com.bok.auth.entity.User;
import com.bok.auth.repository.RefreshTokenRepository;
import com.bok.auth.repository.UserRepository;
import com.bok.auth.security.JwtService;
import com.bok.auth.exception.UserNotFoundException;
import com.bok.auth.dto.AuthResponse;
import com.bok.auth.dto.LoginRequest;
import com.bok.auth.dto.RegisterRequest;
import com.bok.auth.dto.UserResponse;
import com.bok.auth.exception.DuplicationEmailException;
import com.bok.auth.exception.InvalidCredentialsException;
import com.bok.auth.exception.InvalidRefreshTokenException;

import org.springframework.security.crypto.password.PasswordEncoder;

import jakarta.transaction.Transactional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private static final long REFRESH_TOKEN_EXPIRATION_DAYS = 7;

    public AuthService(UserRepository userRepository, RefreshTokenRepository refreshTokenRepository,
                        PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.userRepository = userRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    public AuthResponse login(LoginRequest loginRequest) {
        User user = userRepository.findByEmail(loginRequest.getEmail())
                .orElseThrow(() -> new InvalidCredentialsException());

        if (!passwordEncoder.matches(loginRequest.getPasswordHash(), user.getPasswordHash())) {
            throw new InvalidCredentialsException();
        }
        return buildAuthResponse(user);
    }

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new DuplicationEmailException(request.getEmail());
        }

        User newUser = new User();
        newUser.setFirstName(request.getFirstName());
        newUser.setLastName(request.getLastName());
        newUser.setEmail(request.getEmail());
        newUser.setPhoneNumber(request.getPhoneNumber());
        newUser.setPasswordHash(passwordEncoder.encode(request.getPasswordHash()));
        User saved = userRepository.save(newUser);

        return buildAuthResponse(saved);
    }

    @Transactional
    public AuthResponse refreshAccessToken(String refreshTokenValue) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenValue)
                .orElseThrow(() -> new InvalidRefreshTokenException());

        if (refreshToken.isRevoked() || refreshToken.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new InvalidRefreshTokenException();
        }

        User user = refreshToken.getUser();
        String newAccessToken = jwtService.generateAccessToken(user.getId(), user.getEmail(), user.getRole().name());
        return new AuthResponse(newAccessToken, refreshTokenValue, UserResponse.from(user));
    }

    private AuthResponse buildAuthResponse(User user) {
        String accessToken = jwtService.generateAccessToken(user.getId(), user.getEmail(), user.getRole().name());

        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setUser(user);
        refreshToken.setToken(UUID.randomUUID().toString());
        refreshToken.setExpiresAt(LocalDateTime.now().plusDays(REFRESH_TOKEN_EXPIRATION_DAYS));
        refreshTokenRepository.save(refreshToken);

        return new AuthResponse(accessToken, refreshToken.getToken(), UserResponse.from(user));
    }

    public void logout(String refreshTokenValue) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenValue)
                .orElseThrow(() -> new InvalidRefreshTokenException());
        refreshToken.setRevoked(true);
        refreshTokenRepository.save(refreshToken);
    }

    public User getUserById(UUID id) {
        return userRepository.findById(id).orElseThrow(() -> new UserNotFoundException());
    }

    public List<User> listUsers() {
        return userRepository.findAll();
    }

    public void deleteUser(UUID id) {
        userRepository.deleteById(id);
    }
}

File: Frontend/src/services/authService.ts
import api from "./api";
import type { LoginRequest, AuthResponse, RegisterRequest, UpdatePasswordRequest } from "../types/auth";

export async function login(data: LoginRequest): Promise<AuthResponse> {
  const res = await api.post<AuthResponse>("/api/auth/login", data);
  return res.data;
}

export async function register(data: RegisterRequest): Promise<AuthResponse> {
  const res = await api.post<AuthResponse>("/api/auth/register", data);
  return res.data;
}

export async function logout(refreshToken: string): Promise<void> {
  await api.post("/api/auth/logout", { refreshToken });
}

export async function updatePassword(data: UpdatePasswordRequest): Promise<UpdatePasswordRequest> {
  const res = await api.post<UpdatePasswordRequest>("/api/auth/update-password", data);
  return res.data;
}

export async function refreshAccessToken(): Promise<string | null> {
  const refreshToken = localStorage.getItem("refreshToken");
  if (!refreshToken) return null;
  try {
    const res = await api.post<{ accessToken: string }>("/api/auth/refresh", { refreshToken });
    localStorage.setItem("accessToken", res.data.accessToken);
    return res.data.accessToken;
  } catch {
    return null;
  }
}

export async function getInformation(id: string) : Promise<AuthResponse | null> {
  try {
    const res = await api.get<AuthResponse>(`/api/users/${id}`);
    return res.data;
  } catch {
    return null;
  }
}